smart_contracts/contracts/explorer/faucet-stored/src/main.rs

```diff
@@ -3,57 +3,245 @@
 
 extern crate alloc;
 
-use alloc::{string::ToString, vec};
+use alloc::{boxed::Box, format, string::ToString, vec};
 
-use casper_contract::contract_api::{runtime, storage};
+use casper_contract::{
+    contract_api::{account, runtime, storage, system},
+    unwrap_or_revert::UnwrapOrRevert,
+};
 use casper_types::{
-    account::AccountHash, contracts::ContractHash, CLType, CLTyped, ContractVersion, EntryPoint,
-    EntryPointAccess, EntryPointType, EntryPoints, Parameter,
+    contracts::NamedKeys, ApiError, CLType, EntryPoint, EntryPointAccess, EntryPointType,
+    EntryPoints, Parameter, PublicKey, URef, U512,
 };
 
-const CONTRACT_NAME: &str = "faucet";
-const HASH_KEY_NAME: &str = "faucet_package";
-const ACCESS_KEY_NAME: &str = "faucet_package_access";
-const ENTRY_POINT_NAME: &str = "call_faucet";
-const ARG_TARGET: &str = "target";
-const ARG_AMOUNT: &str = "amount";
-const CONTRACT_VERSION: &str = "contract_version";
+#[repr(u16)]
+enum InstallerSessionError {
+    FailedToTransfer = 101,
+}
 
 #[no_mangle]
 pub extern "C" fn call_faucet() {
     faucet::delegate();
 }
 
-fn store() -> (ContractHash, ContractVersion) {
+fn build_named_keys_and_purse() -> (NamedKeys, URef) {
+    let mut named_keys = NamedKeys::new();
+    let purse = system::create_purse();
+
+    // This session constructs a NamedKeys struct and later passes it to the
+    // storage::new_contract() function. This is simpler and more efficient than creating a custom
+    // "init" entry point for the stored contract in this case but it is not the best approach for
```

Comment:
I'd like to see an explanation as to *why* using the `init` entrypoint pattern here isn't desirable, unless I've completely misunderstood the what is happening here. One can setup the installer to call `init` directly once the contract object has been created, thus allowing other entry points to function if they're depending on sentinel values that are present in the named-keys.

Reply:
The comment gives the rationale: simpler and more efficient.  Let's discuss offline if you feel that's not accurate.

Reply:
The comment gives the rationale: simpler and more efficient.  Let's discuss offline if you feel that's not accurate.